File: lib/services/lifras_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart';
import '../models/exercice_lifras.dart';

/// Service de gestion des exercices LIFRAS
class LifrasService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  /// Récupérer tous les exercices d'un club
  Future<List<ExerciceLIFRAS>> getAllExercices(String clubId) async {
    try {
      final snapshot = await _firestore
          .collection('clubs/$clubId/exercices_lifras')
          .orderBy('code')
          .get();

      final exercices = snapshot.docs
          .map((doc) => ExerciceLIFRAS.fromFirestore(doc))
          .toList();

      debugPrint('📚 ${exercices.length} exercices LIFRAS chargés');
      return exercices;
    } catch (e) {
      debugPrint('❌ Erreur chargement exercices LIFRAS: $e');
      return [];
    }
  }

  /// Récupérer les exercices d'un niveau spécifique + exercices "Tous Niveaux" (TN)
  /// Règle: exercices au niveau exact du membre + tous les exercices TN
  Future<List<ExerciceLIFRAS>> getExercicesByNiveau(
    String clubId,
    NiveauLIFRAS niveau,
  ) async {
    try {
      debugPrint('🔍 Recherche exercices pour niveau: ${niveau.code} (${niveau.label})');

      // Strategy: load all exercises and filter client-side.
      // This avoids composite index requirements (niveau + code)
      // and handles both 'P3' and '3' formats stored in Firestore.
      final allExercices = await getAllExercices(clubId);

      if (allExercices.isEmpty) {
        debugPrint('⚠️ Aucun exercice LIFRAS trouvé dans le club $clubId');
        return [];
      }

      final exercices = <ExerciceLIFRAS>[];

      // 1. Get TN (Tous Niveaux) exercises - accessible to everyone
      final tnExercices = allExercices.where((ex) => ex.niveau == NiveauLIFRAS.tn).toList();
      exercices.addAll(tnExercices);
      debugPrint('📚 ${tnExercices.length} exercices TN (Tous Niveaux) trouvés');

      // 2. Get exercises for the specific niveau (if not TN)
      if (niveau != NiveauLIFRAS.tn) {
        final niveauExercices = allExercices.where((ex) => ex.niveau == niveau).toList();
        exercices.addAll(niveauExercices);
        debugPrint('📚 ${niveauExercices.length} exercices pour niveau ${niveau.label}');
      }

      // Sort by code for consistent display
      exercices.sort((a, b) => a.code.compareTo(b.code));

      debugPrint('📚 ${exercices.length} exercices LIFRAS total pour niveau ${niveau.label}');
      return exercices;
    } catch (e) {
      debugPrint('❌ Erreur chargement exercices par niveau: $e');
      return [];
    }
  }

  /// Récupérer un exercice par son ID
  Future<ExerciceLIFRAS?> getExerciceById(String clubId, String exerciceId) async {
    try {
      final doc = await _firestore
          .collection('clubs/$clubId/exercices_lifras')
          .doc(exerciceId)
          .get();

      if (!doc.exists) {
        debugPrint('❌ Exercice $exerciceId non trouvé');
        return null;
      }

      return ExerciceLIFRAS.fromFirestore(doc);
    } catch (e) {
      debugPrint('❌ Erreur récupération exercice: $e');
      return null;
    }
  }

  /// Récupérer plusieurs exercices par leurs IDs
  Future<List<ExerciceLIFRAS>> getExercicesByIds(
    String clubId,
    List<String> exerciceIds,
  ) async {
    if (exerciceIds.isEmpty) return [];

    try {
      final exercices = <ExerciceLIFRAS>[];

      // Firestore 'in' queries are limited to 10 items, so we batch
      const batchSize = 10;
      for (var i = 0; i < exerciceIds.length; i += batchSize) {
        final batch = exerciceIds.skip(i).take(batchSize).toList();

        final snapshot = await _firestore
            .collection('clubs/$clubId/exercices_lifras')
            .where(FieldPath.documentId, whereIn: batch)
            .get();

        exercices.addAll(
          snapshot.docs.map((doc) => ExerciceLIFRAS.fromFirestore(doc)),
        );
      }

      debugPrint('📚 ${exercices.length}/${exerciceIds.length} exercices LIFRAS récupérés');
      return exercices;
    } catch (e) {
      debugPrint('❌ Erreur récupération exercices par IDs: $e');
      return [];
    }
  }
}

File: src/services/lifrasService.ts
import { collection, doc, getDocs, setDoc, deleteDoc } from 'firebase/firestore';
import { db } from '@/lib/firebase';
import { ExerciceLIFRAS, NiveauLIFRAS } from '@/types/lifras.types';
import { logger } from '@/utils/logger';

const COLLECTION_NAME = 'exercices_lifras';

export const lifrasService = {
  // Récupérer tous les exercices
  async getAllExercices(clubId: string): Promise<ExerciceLIFRAS[]> {
    try {
      const exercicesRef = collection(db, 'clubs', clubId, COLLECTION_NAME);
      const snapshot = await getDocs(exercicesRef);

      return snapshot.docs.map(doc => ({
        id: doc.id,
        ...doc.data(),
        created_at: doc.data().created_at?.toDate(),
        updated_at: doc.data().updated_at?.toDate()
      } as ExerciceLIFRAS));
    } catch (error) {
      logger.error('Erreur lors de la récupération des exercices:', error);
      return [];
    }
  },

  // Récupérer les exercices par niveau
  // Loads all exercises and filters client-side to avoid composite index requirements
  // and to handle both 'P3' and '3' format stored in Firestore
  async getExercicesByNiveau(clubId: string, niveau: NiveauLIFRAS): Promise<ExerciceLIFRAS[]> {
    try {
      const allExercices = await this.getAllExercices(clubId);

      // Filter for the requested niveau + TN (Tous Niveaux) exercises
      const filtered = allExercices.filter(ex =>
        ex.niveau === niveau || ex.niveau === 'TN'
      );

      // Sort by code for consistent display
      filtered.sort((a, b) => a.code.localeCompare(b.code));

      return filtered;
    } catch (error) {
      logger.error(`Erreur lors de la récupération des exercices pour le niveau ${niveau}:`, error);
      return [];
    }
  },

  // Créer ou mettre à jour un exercice
  async saveExercice(clubId: string, exercice: Omit<ExerciceLIFRAS, 'id'>, exerciceId?: string): Promise<string> {
    try {
      const id = exerciceId || doc(collection(db, 'clubs', clubId, COLLECTION_NAME)).id;
      const exerciceRef = doc(db, 'clubs', clubId, COLLECTION_NAME, id);

      // Filter out undefined values (Firestore rejects them)
      const data: Record<string, any> = {
        code: exercice.code,
        niveau: exercice.niveau,
        description: exercice.description,
        updated_at: new Date(),
        created_at: exercice.created_at || new Date(),
      };
      if (exercice.specialite) {
        data.specialite = exercice.specialite;
      }

      await setDoc(exerciceRef, data);

      return id;
    } catch (error) {
      logger.error('Erreur lors de la sauvegarde de l\'exercice:', error);
      throw error;
    }
  },

  // Supprimer un exercice
  async deleteExercice(clubId: string, exerciceId: string): Promise<void> {
    try {
      const exerciceRef = doc(db, 'clubs', clubId, COLLECTION_NAME, exerciceId);
      await deleteDoc(exerciceRef);
    } catch (error) {
      logger.error('Erreur lors de la suppression de l\'exercice:', error);
      throw error;
    }
  },

  // Import en masse des exercices
  async importExercices(clubId: string, exercices: Omit<ExerciceLIFRAS, 'id'>[]): Promise<void> {
    try {
      const promises = exercices.map(exercice =>
        this.saveExercice(clubId, exercice)
      );
      await Promise.all(promises);
    } catch (error) {
      logger.error('Erreur lors de l\'import des exercices:', error);
      throw error;
    }
  },

  // Récupérer la liste des spécialités existantes (pour les exercices TN)
  async getSpecialites(clubId: string): Promise<string[]> {
    try {
      const exercices = await this.getAllExercices(clubId);
      const specialites = exercices
        .filter(ex => ex.niveau === 'TN' && ex.specialite)
        .map(ex => ex.specialite as string);
      return [...new Set(specialites)].sort();
    } catch (error) {
      logger.error('Erreur lors de la récupération des spécialités:', error);
      return [];
    }
  }
};
